Remotion/Web/UnitTests/Core/Context/HttpContextStorageProviderWithAsyncLocalBasedFallbackTest.cs

```diff
@@ -0,0 +1,161 @@
+    }
+
+    [Test]
+    public void SafeContextBoundaryDataUnavailableAfterDispose ()
+    {
+      using (_provider.OpenSafeContextBoundary())
+      {
+        _provider.SetData("test", 123);
+        var safeContextResult = _provider.GetData("test");
+        Assert.That(safeContextResult, Is.EqualTo(123));
+      }
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.Null);
+    }
+
+    [Test]
+    public void SafeContextBoundaryCreationClearsData ()
+    {
+      _provider.SetData("test", 123);
+      _provider.OpenSafeContextBoundary();
+
+      Assert.That(HttpContext.Current.Items["Rm.CtxP"], Is.Null);
+    }
+
+    [Test]
+    public void HttpContextSwitchOnSameThreadResetsData ()
+    {
+      _provider.SetData("test", 123);
+
+      HttpContext.Current = HttpContextHelper.CreateHttpContext("a", "b", "c");
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.Null);
+    }
+
+    [Test]
+    public void HttpContextSharedToNewThreadContainsSameData ()
+    {
+      _provider.SetData("test", 123);
+
+      var context = HttpContext.Current;
+      int? threadResult = null;
+
+      var t = new Thread(
+          () =>
+          {
+            HttpContext.Current = context;
+            threadResult = (int)_provider.GetData("test");
+          });
+      t.Start();
+      t.Join();
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.EqualTo(123));
+      Assert.That(threadResult, Is.EqualTo(123));
+    }
+
+    [Test]
+    public void NewThreadCantAccessExistingData ()
+    {
+      _provider.SetData("test", 123);
+
+      int? threadResult = null;
+
+      var t = new Thread(
+          () =>
+          {
+            threadResult = (int)_provider.GetData("test");
+          });
+      t.Start();
+      t.Join();
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.EqualTo(123));
+      Assert.That(threadResult, Is.Null);
+    }
+
+    [Test]
+    public void HashtableIsCollectedAfterRemovalFromHttpContextItems ()
+    {
+      _provider.SetData("test", 123);
+
+      var hashtable = GetHashtableWeakReference();
+
+      Assert.That(hashtable.IsAlive, Is.True);
+
+      HttpContext.Current.Items["Rm.CtxP"] = null;
+
+      GC.Collect();
+      GC.WaitForPendingFinalizers();
+      GC.Collect();
+
+      Assert.That(hashtable.IsAlive, Is.False);
+
+      WeakReference GetHashtableWeakReference ()
+      {
+        var accessorFieldInfo = typeof(HttpContextStorageProviderWithAsyncLocalBasedFallback).GetField("_hashtableAccessor", BindingFlags.Instance | BindingFlags.NonPublic);
+        var getOrCreateHashtableMethodInfo = accessorFieldInfo!.FieldType.GetMethod("GetOrCreateHashtable");
+        var accessorField = accessorFieldInfo.GetValue(_provider);
+
+        var hashtableFromField = (Hashtable)getOrCreateHashtableMethodInfo!.Invoke(accessorField, new[] { HttpContext.Current });
+
+        return new(hashtableFromField);
```

Comment:
I would suggest we simplify / abstract this a bit:

Return a new WeakReference(new object)), place the WeakReference.Value (=the object) as value into the provider.SetData(...) and then just Assert that this object has been collected during GC. 

The intention of the test would remain the same (= we test that the values aren't kept alive on the thread) but we decouple ourselfs from the implementation detail.